pkg/rotator/rotator.go

```diff
@@ -0,0 +1,536 @@
+package rotator
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	certName               = "tls.crt"
+	keyName                = "tls.key"
+	caCertName             = "ca.crt"
+	caKeyName              = "ca.key"
+	rotationCheckFrequency = 12 * time.Hour
+	certValidityDuration   = 10 * 365 * 24 * time.Hour
+)
+
+var crLog = logf.Log.WithName("cert-rotation")
+
+var vwhGVK = schema.GroupVersionKind{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "ValidatingWebhookConfiguration"}
+
+var _ manager.Runnable = &CertRotator{}
+
+// AddRotator adds the CertRotator and ReconcileVWH to the manager.
+func AddRotator(mgr manager.Manager, cr *CertRotator, vwhName string) error {
```

Comment:
Not sure if we want to address this in the initial commit or not but: what are the cardinalities we have in mind? Eg, suppose we want to run N validating/mutating webhooks. How many CertRotators does that map to? 1, N, or ...? Also how many webhook controller/reconciler processes?

Reply:
The original idea is that all pods that participate in a webhook are running the cert-controller code.

This creates a worst-case scenario of last-write wins, and has the advantage that the pods don't need to coordinate with each other directly and no external management process is needed.

In practice, I think the granularity is at the secret level:

* All pods that are part of the informal "last write wins" ring run the controller against the secret.
* That secret populates all registered webhooks in a single VWH config (not strictly necessary, but simpler than trying to carve out individual webhooks)
* A single secret can fan out to as many VWH configs as we like. Different pods can fan out to different VWH configs.